packages/shared/src/porter.ts

```diff
@@ -284,4 +354,88 @@ export class PorterClient {
     > = Object.fromEntries(decryptionResponses);
     return { encryptedResponses, errors };
   }
-}
+
+
+  public async signUserOp(
+    signingRequests: Record<string, string>,
+    threshold: number,
+    options: SigningOptions = {},
+  ): Promise<SignResult> {
+    const data: Record<string, unknown> = {
+      signing_requests: signingRequests,
+      threshold: threshold
+    };
+    
+    const resp: AxiosResponse<SignResponse> = await this.tryAndCall({
+      url: '/sign',
+      method: 'post',
+      data,
+    });
+
+    const { signatures, errors } = resp.data.result.signing_results;
+    const allErrors: Record<string, string> = { ...errors };
+    const { optimistic = false } = options;
+    
+    const signingResults: { [ursulaAddress: string]: [string, string] } = {};
+    let messageHash = '';
+    
+    // For non-optimistic path
+    const hashCounts: Map<string, number> = new Map();
+    const hashToSignatures: Map<string, { [ursulaAddress: string]: [string, string] }> = new Map();
+    
+    // Single pass: decode signatures and populate signingResults
+    for (const [ursulaAddress, [signerAddress, signatureB64]] of Object.entries(signatures || {})) {
+      const decoded = decodeSignature(ursulaAddress, signerAddress, signatureB64);
+      if (decoded.error) {
+        allErrors[ursulaAddress] = decoded.error;
+        continue;
+      }
+      
+      if (!messageHash) {
+        messageHash = decoded.result!.messageHash;
+      }
+      
+      // Always include all decoded signatures in signingResults
+      signingResults[ursulaAddress] = [decoded.result!.signerAddress, decoded.result!.signature];
+      
+      if (!optimistic) {
+        // For non-optimistic: track hashes and group signatures for aggregation
+        const hash = decoded.result!.messageHash;
+        hashCounts.set(hash, (hashCounts.get(hash) || 0) + 1);
+        
+        if (!hashToSignatures.has(hash)) {
+          hashToSignatures.set(hash, {});
+        }
+        hashToSignatures.get(hash)![ursulaAddress] = [decoded.result!.signerAddress, decoded.result!.signature];
+      }
+    }
+
+    // Determine which signatures to aggregate
+    let signaturesToAggregate: { [ursulaAddress: string]: [string, string] } = {};
+    
+    if (optimistic) {
+      // Optimistic: aggregate all signatures
+      signaturesToAggregate = signingResults;
+    } else {
+      // Non-optimistic: only aggregate signatures from hash that meets threshold
+      for (const [hash, count] of hashCounts.entries()) {
+        if (count >= threshold) {
+          signaturesToAggregate = hashToSignatures.get(hash)!;
```

Comment:
Do you need to update the messageHash var here?
```suggestion
          messageHash = hash
          signaturesToAggregate = hashToSignatures.get(hash)!;
```

Reply:
✔️